packages/asset-swapper/src/utils/quote_requestor.ts

```diff
@@ -0,0 +1,99 @@
+import { assetDataUtils, SignedOrder } from '@0x/order-utils';
+import { ERC20AssetData } from '@0x/types';
+import { BigNumber, logUtils } from '@0x/utils';
+import Axios, { AxiosResponse } from 'axios';
+
+import { MarketOperation } from '../types';
+
+/**
+ * Request quotes from RFQ-T providers
+ */
+export class QuoteRequestor {
+    private readonly _rfqtMakerEndpoints: string[];
+
+    constructor(rfqtMakerEndpoints: string[]) {
+        this._rfqtMakerEndpoints = rfqtMakerEndpoints;
+    }
+
+    public async requestRfqtFirmQuotesAsync(
+        makerAssetData: string,
+        takerAssetData: string,
+        assetFillAmount: BigNumber,
+        marketOperation: MarketOperation,
+        intentOnFilling: boolean,
+        takerApiKey: string,
+        takerAddress: string,
+    ): Promise<SignedOrder[]> {
+        const makerEndpointMaxResponseTimeMs = 1000;
+
+        const getTokenAddressOrThrow = (assetData: string): string => {
+            const decodedAssetData = assetDataUtils.decodeAssetDataOrThrow(assetData);
+            if (decodedAssetData.hasOwnProperty('tokenAddress')) {
+                // type cast necessary here as decodeAssetDataOrThrow returns
+                // an AssetData object, which doesn't necessarily contain a
+                // token address.  (it could possibly be a StaticCallAssetData,
+                // which lacks an address.)  so we'll just assume it's a token
+                // here.  should be safe, with the enclosing guard condition
+                // and subsequent error.
+                // tslint:disable-next-line:no-unnecessary-type-assertion
+                return (decodedAssetData as ERC20AssetData).tokenAddress;
+            } else {
+                throw new Error(
+                    `Decoded asset data (${JSON.stringify(decodedAssetData)}) does not contain a token address`,
+                );
+            }
+        };
+
+        const buyToken = getTokenAddressOrThrow(makerAssetData);
+        const sellToken = getTokenAddressOrThrow(takerAssetData);
+
+        // create an array of promises for quote responses, using "undefined"
+        // as a placeholder for failed requests.
+
+        const responsePromises: Array<Promise<AxiosResponse<SignedOrder> | undefined>> = [];
+        for (const rfqtMakerEndpoint of this._rfqtMakerEndpoints) {
+            responsePromises.push(
```

Comment:
Nit: We can maybe do this in a more idomatic fashion with an async lambda like:

```ts
const responsesIfDefined = await Promise.all(this._rfqtMakerEndpoints.map(async endpoint => {
    try {
        return await Axios.get(...);
    } catch (err) {
        ...
    }
}));
```

TS should also be able to infer the return type this way also.

Reply:
Addressed in https://github.com/0xProject/0x-monorepo/pull/2541/commits/98fc780ade384f732bb71c0e62f77d5bec769fb9 .  Is that what you had in mind?

Reply:
Partly. Does it not work with `try/catch`?

Reply:
Addressed in https://github.com/0xProject/0x-monorepo/pull/2541/commits/121d51b414e698d94eb8d1a3a9d3130ff7431dc3 . Was there anything else I missed?